src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3659,8 +3659,399 @@ class StubGenerator: public StubCodeGenerator {
+      const int qr_end = vset_sew == Assembler::e32 ? 12 : 16;
+      for (int i = 0; i < qr_end; i++) {
+        sha2_quad_round(vset_sew,
+                   rotation_regs[(rot_pos + 0) & 0x3],
+                   rotation_regs[(rot_pos + 1) & 0x3],
+                   rotation_regs[(rot_pos + 2) & 0x3],
+                   rotation_regs[(rot_pos + 3) & 0x3],
+                   consts,
+                   v15, v14, v16, v17);
+        ++rot_pos;
+      }
+      // Quad-round #12 (+0, v10->v11->v12->v13) ... #15 (+3, v13->v10->v11->v12)
+      // Note that we stop generating new message schedule words (Wt, v10-13)
+      // as we already generated all the words we end up consuming (i.e., W[63:60]).
+      const int qr_c_end = qr_end + 4;
+      for (int i = qr_end; i < qr_c_end; i++) {
+        sha2_quad_round(vset_sew,
+                   rotation_regs[(rot_pos + 0) & 0x3],
+                   rotation_regs[(rot_pos + 1) & 0x3],
+                   rotation_regs[(rot_pos + 2) & 0x3],
+                   rotation_regs[(rot_pos + 3) & 0x3],
+                   consts,
+                   v15, v14, v16, v17, false, i < (qr_c_end-1));
+        ++rot_pos;
+      }
+
+      //--------------------------------------------------------------------------------
+      // Compute the updated hash value H'
+      //   H' = H + {h',g',...,b',a'}
+      //      = {h,g,...,b,a} + {h',g',...,b',a'}
+      //      = {h+h',g+g',...,b+b',a+a'}
+
+      // H' = H+{a',b',c',...,h'}
+      __ vadd_vv(v16, v26, v16);
+      __ vadd_vv(v17, v27, v17);
+
+      if (multi_block) {
+        int total_adds = vset_sew == Assembler::e32 ? 240 : 608;
+        __ addi(consts, consts, -total_adds);
+        __ add(ofs, ofs, vset_sew == Assembler::e32 ? 64 : 128);
+        __ ble(ofs, limit, multi_block_loop);
+        __ mv(c_rarg0, ofs); // return ofs
+      }
+
+      // Store H[0..8] = {a,b,c,d,e,f,g,h} from
+      //  v16 = {f,e,b,a}
+      //  v17 = {h,g,d,c}
+      __ vid_v(v30);                                   // v30 = {3,2,1,0}
+      __ vxor_vi(v30, v30, 0x3);                       // v30 = {0,1,2,3}
+      __ vrgather_vv(v26, v16, v30);                   // v26 = {f,e,b,a}
+      __ vrgather_vv(v27, v17, v30);                   // v27 = {h,g,d,c}
+      __ vmsgeu_vi(v0, v30, 2);                        // v0  = {f,f,t,t}
+      // Copy elements [3..2] of v26 ({f,e}) into elements [1..0] of v17.
+      __ vslidedown_vi(v17, v26, 2);                   // v17 = {_,_,f,e}
+      // Merge elements [3..2] of v27 ({g,h}) into elements [3..2] of v17
+      __ vmerge_vvm(v17, v27, v17);                    // v17 = {h,g,f,e}
+      // Copy elements [1..0] of v27 ({c,d}) into elements [3..2] of v16.
+      __ vslideup_vi(v16, v27, 2);                     // v16 = {d,c,_,_}
+      // Merge elements [1..0] of v26 ({a,b}) into elements [1..0] of v16
+      __ vmerge_vvm(v16, v16, v26);                    // v16 = {d,c,b,a}
```

Comment:
Great to see that we are switching to use index-load to get {f,e,b,a},{h,g,d,c} pre-loop. But I was also expecting to use index-store to put {f,e,b,a},{h,g,d,c} back to {a,b,c,d},{e,f,g,h} post-loop as I mentioned in my previous comment [1]. Seems that was missed?

[1] https://github.com/openjdk/jdk/pull/16562#discussion_r1393767555

Reply:
Yes, I missed that, thanks!

Reply:
Fixed